File: Linked List/234.palindrome-linked-list.py
#
# @lc app=leetcode id=234 lang=python3
#
# [234] Palindrome Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True
        slow, fast, pre, prepre = head, head, head, None
        while fast and fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next
            pre.next = prepre
            prepre = pre
        if fast:
            slow = slow.next
        while pre and slow:
            if pre.val != slow.val:
                return False
            pre = pre.next
            slow = slow.next
        return True
# @lc code=end

File: Linked List/234.palindrome-linked-list.cpp
/*
 * @lc app=leetcode id=234 lang=cpp
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head) return true;
        auto p = head;
        int size = 1;
        while(p->next) {
            p = p->next;
            size++;
        }
        auto fast = head->next;
        auto slow = head;
        while(fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }
        slow = slow->next;
        auto cur = head;
        ListNode* prev = nullptr;
        while(cur != slow) {
            auto next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        if(size % 2 != 0) prev = prev->next;
        while(prev && slow) {
            if(slow->val != prev->val) return false;
            prev = prev->next;
            slow = slow->next;
        }
        return true;
    }
};
// @lc code=end

